lib/src/native/realm_core.dart

```diff
@@ -554,67 +553,79 @@ class _RealmCore {
     return _realmLib.realm_list_is_valid(list.handle._pointer);
   }
 
-  static void collection_change_callback(Object object, Pointer<realm_collection_changes> data) {
-    assert(object is NotificationsController, "Notification controller expected");
-
-    final controller = object as NotificationsController;
-
-    if (data == nullptr) {
-      //realm_collection_changes data clone is done in native code before this callback is invoked. nullptr data means cloning failed.
-      controller.onError(RealmError("Invalid notifications data received"));
-      return;
-    }
+  static void collection_change_callback(Pointer<Void> userdata, Pointer<realm_collection_changes> data) {
+    final controller = _realmLib.gc_handle_deref(userdata);
+    if (controller is NotificationsController) {
+      if (data == nullptr) {
+        controller.onError(RealmError("Invalid notifications data received"));
+        return;
+      }
 
-    try {
-      final changesHandle = RealmCollectionChangesHandle._(data);
-      controller.onChanges(changesHandle);
-    } catch (e) {
-      controller.onError(RealmError("Error handling collection change notifications. Error: $e"));
+      try {
+        final changesHandle = RealmCollectionChangesHandle._(_realmLib.realm_clone(data.cast()).cast());
+        controller.onChanges(changesHandle);
+      } catch (e) {
+        controller.onError(RealmError("Error handling collection change notifications. Error: $e"));
+      }
     }
   }
 
-  static void object_change_callback(Object object, Pointer<realm_object_changes> data) {
-    assert(object is NotificationsController, "Notification controller expected");
-
-    final controller = object as NotificationsController;
-
-    if (data == nullptr) {
-      //realm_collection_changes data clone is done in native code before this callback is invoked. nullptr data means cloning failed.
-      controller.onError(RealmError("Invalid notifications data received"));
-      return;
-    }
+  static void object_change_callback(Pointer<Void> userdata, Pointer<realm_object_changes> data) {
+    final controller = _realmLib.gc_handle_deref(userdata);
+    if (controller is NotificationsController) {
+      if (data == nullptr) {
+        //realm_collection_changes data clone is done in native code before this callback is invoked. nullptr data means cloning failed.
+        controller.onError(RealmError("Invalid notifications data received"));
+        return;
+      }
 
-    try {
-      final changesHandle = RealmObjectChangesHandle._(data);
-      controller.onChanges(changesHandle);
-    } catch (e) {
-      controller.onError(RealmError("Error handling collection change notifications. Error: $e"));
+      try {
+        final changesHandle = RealmObjectChangesHandle._(_realmLib.realm_clone(data.cast()).cast());
```

Comment:
realm_clone should be invoked with tryGetPointer and could return `null` which we should check